File: GoodBook/InputView.m
//
//  InputView.m
//  GoodBook
//
//  Created by zhy on 17/4/6.
//  Copyright © 2017年 zhanghaiyong. All rights reserved.
//

#import "InputView.h"

@implementation InputView

-(void)awakeFromNib {

    [super awakeFromNib];
    
    self.textView.delegate = self;
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyBoardWillShow:) name:UIKeyboardWillShowNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyBoardWillHide:) name:UIKeyboardWillHideNotification object:nil];
}

- (void)keyBoardWillShow:(NSNotification *)notification {

    //键盘frame
    CGRect keyBoardFrame;
    [notification.userInfo[UIKeyboardFrameEndUserInfoKey] getValue:&keyBoardFrame];
    
    //键盘显示时间
    double duration;
    [notification.userInfo[UIKeyboardAnimationDurationUserInfoKey] getValue:&duration];
    
    //键盘动画
    UIViewAnimationCurve curve;
    [notification.userInfo[UIKeyboardAnimationCurveUserInfoKey] getValue:&curve];
    
    
    if ([self.delegate respondsToSelector:@selector(keyBoardWillShow:keyBoardH:animationDuration:animationCurve:)]) {
        
        [self.delegate keyBoardWillShow:self keyBoardH:keyBoardFrame.size.height animationDuration:duration animationCurve:curve];
    }
}

- (void)keyBoardWillHide:(NSNotification *)notification {
    
    //键盘frame
    CGRect keyBoardFrame;
    [notification.userInfo[UIKeyboardFrameEndUserInfoKey] getValue:&keyBoardFrame];
    
    //键盘显示时间
    double duration;
    [notification.userInfo[UIKeyboardAnimationDurationUserInfoKey] getValue:&duration];
    
    //键盘动画
    UIViewAnimationCurve curve;
    [notification.userInfo[UIKeyboardAnimationCurveUserInfoKey] getValue:&curve];
    
    if ([self.delegate respondsToSelector:@selector(keyBoardWillHide:keyBoardH:animationDuration:animationCurve:)]) {
        
        [self.delegate keyBoardWillHide:self keyBoardH:keyBoardFrame.size.height animationDuration:duration animationCurve:curve];
    }
}

#pragma mark UITextViewDelegate
- (BOOL)textView:(UITextView *)textView shouldChangeTextInRange:(NSRange)range replacementText:(NSString *)text
{
    NSMutableString *times = [[NSMutableString alloc]initWithFormat:@"%@", textView.text];
    //字符串查找,可以判断字符串中是否有
    if ([times hasPrefix:@"@"]) {
        textView.text = @"";
        // textView.textColor =  RGB(70, 70, 70);
    }
    return YES;
}

- (void)textViewDidChange:(UITextView *)textView {

    CGRect frame = [self.textView.text boundingRectWithSize:CGSizeMake(self.textView.frame.size.width, MAXFLOAT) options:NSStringDrawingUsesFontLeading | NSStringDrawingUsesLineFragmentOrigin attributes:@{NSFontAttributeName:self.textView.font} context:nil];
    
        if ([self.delegate respondsToSelector:@selector(textViewHeightDidChange:)]) {
            
            [self.delegate textViewHeightDidChange:frame.size.height];
        }
}

- (IBAction)releaseAction:(id)sender {
    
    if ([self.delegate respondsToSelector:@selector(releaseComment:)]) {
        
        [self.delegate releaseComment:self.textView.text];
    }
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

@end

File: GoodBook/InputView.swift
//
//  InputView.swift
//  GoodBook
//
//  Created by zhy on 17/4/1.
//  Copyright © 2017年 zhanghaiyong. All rights reserved.
//

import UIKit

protocol InputViewDelegate {
    
    func keyboardWillShow(inputView : InputView,keyBoardH : CGFloat,duration : TimeInterval,animation : UIViewAnimationCurve)
    func keyboardWillHide(inputView : InputView,keyBoardH : CGFloat,duration : TimeInterval,animation : UIViewAnimationCurve)
    func textViewHeightDidChange(height : CGFloat)
}

class InputView: UIView,UITextViewDelegate {
    
    @IBOutlet weak var textView: UITextView!
    var delegate : InputViewDelegate?
    
    required init?(coder aDecoder: NSCoder) {
        
        super.init(coder: aDecoder)
        self.textView.delegate = self;
        
        //注册键盘通知
        NotificationCenter.default.addObserver(self, selector: #selector(keyBoardWillShow(notifation:)), name: NSNotification.Name.UIKeyboardWillShow, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(keyBoardWillHide(notifation:)), name: NSNotification.Name.UIKeyboardWillHide, object: nil)
    }
    
    
    @IBAction func subitAction(_ sender: Any) {
    }
    
    
    //键盘通知
    func keyBoardWillShow(notifation : Notification) {
        
        //键盘frame
        let frame = notifation.userInfo?[UIKeyboardFrameEndUserInfoKey] as? CGRect
        let animationCurve = notifation.userInfo?[UIKeyboardAnimationCurveUserInfoKey]
        let duration = notifation.userInfo?[UIKeyboardAnimationDurationUserInfoKey]
        self.delegate?.keyboardWillShow(inputView: self, keyBoardH: (frame?.size.height)!, duration: duration as! TimeInterval, animation: animationCurve as! UIViewAnimationCurve)
        
    }
    
    func keyBoardWillHide(notifation : Notification) {
        
        //键盘frame
        let frame = notifation.userInfo?[UIKeyboardFrameEndUserInfoKey] as? CGRect
        let animationCurve = notifation.userInfo?[UIKeyboardAnimationCurveUserInfoKey]
        let duration = notifation.userInfo?[UIKeyboardAnimationDurationUserInfoKey]
        self.delegate?.keyboardWillHide(inputView: self, keyBoardH: (frame?.size.height)!, duration: duration as! TimeInterval, animation: animationCurve as! UIViewAnimationCurve)
    }
    
    //UITextViewDelegate
    func textViewDidChange(_ textView: UITextView) {
        
//        var rect = self.textView.text.boundingRectWithSize(CGSizeMake(self.frame.size.width-16,CGFloat.max), options: NSStringDrawingOptions.UsesLineFragmentOrigin, attributes: attribute_set as [NSObject : AnyObject], context:nil)
        
    }
}









